### DISCUSSION TITLE
Implementing password login without JS in PocketBase as framework

### DISCUSSION THREAD

**Discussion by mhabala** [pocketbase/pocketbase#2347] [Q&A] [answered]
Hello everyone,

I'm currently using PocketBase as framework and I'm having trouble figuring out how to implement login with a password without using JavaScript.

Currently, I have created a login form, but I'm not sure what my next steps should be. When I submit the form to "/api/collections/users/auth-with-password", I receive a response object. However, my goal is to set a cookie with a token and redirect the user to a different page after they've successfully logged in.

**Comment by ganigeorgiev** [comment]
It is a little outdated, but you can use https://github.com/pocketbase/pocketbase/discussions/989#discussioncomment-4109411 as a reference for loading the auth context from a cookie.

For setting the cookie string, you can [register a new route](https://pocketbase.io/docs/router/#register-new-route) or attach to the [`OnRecordAuthRequest` hook](https://pocketbase.io/docs/event-hooks/#record-auth-api-hooks) and set the JWT token as cookie in the `e.HttpContext`.

**Reply by ganigeorgiev** [reply]
_Additionally, if you need to generate the token manually (eg. in a custom route), you can use the `tokens` subpackage helpers - https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.15.1/tokens._

**Comment by mhabala** [comment]
I have managed to implement loading the auth context from a cookie using the reference provided. However, I am still having trouble registering the route and setting the JWT token as a cookie in the e.HttpContext. Could you please provide an example of implementation of a new route? What I want to know is what logic to use that takes login credentials and gives the token.

**Reply by ganigeorgiev** [reply, marked as answer]
Here is a rough example (_I haven't tested it_):

```go
package main

import (
	"log"
	"net/http"
	"time"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tokens"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		// registes a new "POST /auth" handler
		e.Router.POST("/auth", func(c echo.Context) error {
			data := &struct {
				Email    string `form:"email" json:"email"`
				Password string `form:"password" json:"password"`
			}{}

			// read the request data
			if err := c.Bind(data); err != nil {
				return apis.NewBadRequestError("Failed to read request data", err)
			}

			// fetch the user and check the provided password
			record, err := app.Dao().FindAuthRecordByEmail("users", data.Email)
			if err != nil || !record.ValidatePassword(data.Password) {
				return apis.NewBadRequestError("Invalid login credentials", err)
			}

			// generate a new auth token for the found user record
			token, err := tokens.NewRecordAuthToken(app, record)
			if err != nil {
				return apis.NewBadRequestError("Failed to create an auth token", err)
			}

			// set it as cookie
			cookie := &http.Cookie{
				Name:     "pb_auth", // rename with the name of your cookie
				Value:    token,
				Secure:   true,
				HttpOnly: true,
				SameSite: http.SameSiteStrictMode,
				// you can use the token duration or any other expiration (eg. only 1 day)
				Expires:  time.Now().Add(time.Duration(app.Settings().RecordAuthToken.Duration) * time.Second),
			}

			c.SetCookie(cookie)

			return c.NoContent(http.StatusNoContent)
		}, apis.ActivityLogger(app))

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```
